keps/sig-apps/3939-allow-replacement-when-fully-terminated/README.md

```diff
@@ -0,0 +1,877 @@
+`podReplacementPolicy: Failed`, given the above problems.
+We could even consider removing the proposed field `podReplacementPolicy`.
+
+But for backwards compatibility, in v1, we have to introduce a change of
+behavior as opt-in.
+
+#### When Pods enter a terminating state
+
+Pods can be marked for termination by several controllers, which we typically
+refer to as disruptions, such as: kubelet eviction, scheduler preemption, API eviction, etc.
+
+The job controller itself can delete running Pods, in the following scenarios:
+
+1. A job is over the `activeDeadlineSeconds`.  
+1. When the number of Pod failures reaches the `backoffLimit`.
+1. With `PodFailurePolicy` active and `FailJob` is set as the action.
+
+In all these situations, the Pod initially gets a `deletionTimestamp`
+and we interpret the pod as "terminating". Once the pod terminates, it gets
+a terminal `phase` (`Succeeded` or `Failed`).
+
+### Exponential Backoff for Pod Failures
+
+The job controller implements backoff delays to prevent fast recreation of
+continuously failing Pods.
+
+This behavior is internal (not configurable through the API) and it's orthogonal
+to this KEP. The behavior will be preserved as follows:
+- When `podReplacementPolicy: TerminatingOrFailed`, the backoff period counts from
+  the time the Pod is terminating or Failed.
+- When `podReplacementPolicy: Failed`, the backoff period counts from the time the
+  Pod is Failed.
+
+### Risks and Mitigations
+
+#### Pods are not guaranteed to transition to a terminal phase
+
+One area of contention is how this KEP will work with [3329-retriable-and-non-retriable-failures](https://github.com/kubernetes/enhancements/blob/master/keps/sig-apps/3329-retriable-and-non-retriable-failures/README.md).
+
+In 3329, there was a decision to make kubelet transition pods to failed before deleting them. This is feature toggled guarded by `PodDisruptionCondition`.  
+This means that when this feature is turned on, the job controller is able to count pods as failed only when they are fully terminated, as it is guaranteed that all pods will reach a terminal state (Failed or Succeeded).
+Note that a terminating pod is not considered active either.
+If `PodDisruptionCondition` is turned off, then the job controller considers the pod as failed as soon as it is terminating (has a deletion timestamp), because there is no guarantee that the pod will transition to phase=Failed.
+
+Another issue is described [here](https://github.com/kubernetes/enhancements/pull/3940#discussion_r1180777509).
+If PodDisruptionConditions is disabled, a pod bound to a no-longer-existing node may be stuck in the Running phase.
+As a consequence, it will never be replaced, so the whole job will be stuck from making progress.
+When PodDisruptionConditions is enabled, the PodGC transitions the Pod to phase Failed in this scenario.
+
+Due to the above issues, we propose the following mitigation:
```

Comment:
I'm sorry - these 3 lines are completely unclear to me. Can you somehow rewrite them to make it more clear what exactly you will be doing?

Reply:
This is much better thanks!